talk/objectorientation/adl.tex

```diff
@@ -144,3 +144,27 @@
     \end{columns}
   \end{block}
 \end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[98]{ADL and Hidden Friends}
+  \begin{block}{Hidden friend declarations participate in ADL}
+    \begin{itemize}
+      \item Friends defined in class scope are ``hidden friends''
+      \item They are \emph{not} class members (no access to \mintinline{cpp}{this})
+      \item They participate in ADL, e.g.\ \mintinline{cpp}{std::cout << complex}
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{Ostream operator as hidden friend}
+    \begin{cppcode*}{gobble=2,linenos=false}
+      class Complex {
+        float m_real, m_imag; // + Complex(float, float) ...
+        friend
+        std::ostream & operator<<(std::ostream & os,
+                                  Complex const & c) {
+          return os << c.m_real << " + " << c.m_imag << "i";
+        }
+      };
+      std::cout << Complex(2.f, 3.f); // Prints 2 + 3i
```

Comment:
You could make it an aggregate :)
```suggestion
      struct Complex {
        float m_real, m_imag;
        friend
        std::ostream & operator<<(std::ostream & os,
                                  Complex const & c) {
          return os << c.m_real << " + " << c.m_imag << "i";
        }
      };
      std::cout << Complex{2.f, 3.f}; // Prints 2 + 3i
```
Feel free to ignore this one.

Reply:
I know. But if I do, I don't need to be a friend. 😉

One thing about the `{ }`. I believe that we're not at `{ }` yet. However, these days, we actually want the `{ }`, so should we use them as much as makes sense even before we explain the details?

Reply:
> I know. But if I do, I don't need to be a friend.

Right. Ignore my comment then :)

> One thing about the { }. I believe that we're not at { } yet. However, these days, we actually want the { }, so should we use them as much as makes sense even before we explain the details?

Yes, I would use them.